# Crear Target

Instalamos / actualizamos librerías

In [1]:
##%%bash 
# ! pip install duckdb
# ! pip install jupysql
# ! pip install duckdb-engine
# ! pip install pandas

In [4]:
# ! pip install --upgrade duckdb
# ! pip install --upgrade jupysql
# ! pip install --upgrade duckdb-engine

Configuramos el entorno de ejecución. Si ya tiene todo instalado, solo necesita ejecutar esta celda para empezar a usar **duckdb** 

In [1]:
import duckdb
import pandas as pd

%load_ext sql
%config SqlMagic.autopandas = True
%config SqlMagic.feedback = False
%config SqlMagic.displaycon = False

%sql duckdb:///C:/Users/vanes/Documents/UBA/2do_cuatrimestre/DMEyF/dmeyf.db

Cargamos el archivo `.csv` a una tabla:

In [2]:
%%sql
create or replace table competencia_02 as
select
    *
from read_csv_auto('C:/Users/vanes/Documents/UBA/2do_cuatrimestre/DMEyF/datasets/competencia_02_crudo.csv.gz')


Empty DataFrame
Columns: [Success]
Index: []

Observamos registros para un cliente particular:

In [3]:
%%sql
SELECT 
    * 
FROM competencia_02
WHERE numero_de_cliente = 29215143

numero_de_cliente  foto_mes  active_quarter  cliente_vip  internet  \
0            29215143    201901               1            0         1   
1            29215143    201902               1            0         1   
2            29215143    201903               1            0         1   
3            29215143    201904               1            0         1   
4            29215143    201905               1            0         1   
5            29215143    201906               1            0         1   
6            29215143    201907               1            0         1   
7            29215143    201908               1            0         1   
8            29215143    201909               1            0         1   
9            29215143    201910               1            0         1   
10           29215143    201911               1            0         1   
11           29215143    201912               1            0         1   
12           29215143    202001               1            0         1   
13           29215143    202002               1            0         1   
14           29215143    202003               1            0         1   
15           29215143    202004               1            0         1   
16           29215143    202005               1            0         1   
17           29215143    202006               0            0         0   
18           29215143    202007               1            0         1   
19           29215143    202008               1            0         1   
20           29215143    202009               1            0         1   
21           29215143    202010               1            0         0   
22           29215143    202011               1            0         0   
23           29215143    202012               1            0         0   
24           29215143    202101               1            0         0   
25           29215143    202102               1            0         0   
26           29215143    202103               1            0         0   
27           29215143    202104               1            0         0   
28           29215143    202105               1            0         0   
29           29215143    202106               1            0         0   
30           29215143    202107               1            0         0   

    cliente_edad  cliente_antiguedad  mrentabilidad  mrentabilidad_annual  \
0             55                 257       18365.44             108081.78   
1             55                 258        5537.20             111874.25   
2             55                 259       10687.02             115257.44   
3             55                 260        9781.35             115405.24   
4             55                 261           0.00                  0.00   
5             55                 262       17463.89             128846.58   
6             55                 263       10091.09             133807.98   
7             55                 264        8363.86             138507.59   
8             55                 265        1385.99             135662.45   
9             55                 266           0.00                  0.00   
10            56                 267         849.55             118219.26   
11            56                 268        4221.38              98240.86   
12            56                 269       18930.04              98805.46   
13            56                 270        1166.20              94434.46   
14            56                 271        6555.69              90303.13   
15            56                 272        7907.88              88429.66   
16            56                 273        5936.13              86210.56   
17            56                 274           0.00                  0.00   
18            56                 275       13140.68              85086.76   
19            56                 276       13869.45              90592.36   
20            56  

Basado en sugerencia del profesor se construye target, se incluye lógica para generar la clase_ternaria:

In [4]:
%%sql
create or replace table targets as
with periodos as (
    select distinct foto_mes from competencia_02
), clientes as (
    select distinct numero_de_cliente from competencia_02
), todo as (
    select numero_de_cliente, foto_mes from clientes cross join periodos
), clase_ternaria as (
    select
        t.numero_de_cliente
        , t.foto_mes
        , if(c.numero_de_cliente is null, 0, 1) as mes_0
        , lead(mes_0, 1) over (partition by t.numero_de_cliente order by foto_mes) as mes_1
        , lead(mes_0, 2) over (partition by t.numero_de_cliente order by foto_mes) as mes_2
        , case 
        when (mes_0 = 1) AND (mes_1 = 0) AND  (mes_2 = 0) THEN 'BAJA+1'
        when (mes_0 = 1) AND (mes_1 = 0) AND  (mes_2 = 1) THEN 'CONTINUA'
        when (mes_0 = 1) AND (mes_1 = 1) AND  (mes_2 = 0) THEN 'BAJA+2'
        when (mes_0 = 1) AND (mes_1 = 1) AND  (mes_2 = 1) THEN 'CONTINUA'
        else NULL end as clase_ternaria 
    from todo t
    left join competencia_02 c using (numero_de_cliente, foto_mes)
) select
    foto_mes
  , numero_de_cliente
  , clase_ternaria
from clase_ternaria where mes_0 = 1

Empty DataFrame
Columns: [Success]
Index: []

Algunas verificaciones:

In [31]:
%%sql
SELECT 
    * 
FROM targets
WHERE (mes_0 = 1) AND (mes_1 = 0) AND  (mes_2 = 1)

numero_de_cliente  foto_mes  mes_0  mes_1  mes_2 clase_ternaria
0            30460466    201911      1      0      1       CONTINUA
1           100084897    201912      1      0      1       CONTINUA
2           153254472    202008      1      0      1       CONTINUA
3           109321500    202011      1      0      1       CONTINUA
4           127524491    202005      1      0      1       CONTINUA
5            33496786    201908      1      0      1       CONTINUA
6            76577049    202004      1      0      1       CONTINUA
7           114811302    201902      1      0      1       CONTINUA
8           105229225    202004      1      0      1       CONTINUA
9           116673221    201901      1      0      1       CONTINUA
10           54364396    201908      1      0      1       CONTINUA
11           68212097    201904      1      0      1       CONTINUA
12          126572507    201909      1      0      1       CONTINUA
13          109325590    201910      1      0      1       CONTINUA
14           30012916    201911      1      0      1       CONTINUA
15           67187867    202104      1      0      1       CONTINUA
16          162698675    202105      1      0      1       CONTINUA
17           33677914    201908      1      0      1       CONTINUA
18           62785335    202004      1      0      1       CONTINUA
19          124836780    202001      1      0      1       CONTINUA
20           57311390    201909      1      0      1       CONTINUA
21          123406579    201910      1      0      1       CONTINUA
22           60675670    202003      1      0      1       CONTINUA
23           33148356    201912      1      0      1       CONTINUA
24           35870218    201908      1      0      1       CONTINUA
25           29289549    201911      1      0      1       CONTINUA
26           40950395    201908      1      0      1       CONTINUA
27           32393521    201908      1      0      1       CONTINUA
28           54537778    201906      1      0      1       CONTINUA
29          113011323    202102      1      0      1       CONTINUA
30           55934435    202009      1      0      1       CONTINUA
31          100081285    202007      1      0      1       CONTINUA
32           37223108    202101      1      0      1       CONTINUA
33          106624344    201908      1      0      1       CONTINUA
34          125939345    202006      1      0      1       CONTINUA
35           55190701    201908      1      0      1       CONTINUA
36          170424992    201909      1      0      1       CONTINUA
37           67013204    202104      1      0      1       CONTINUA
38           38970389    202007      1      0      1       CONTINUA
39           89702809    202101      1      0      1       CONTINUA
40           95013285    202102      1      0      1       CONTINUA
41           95380426    202103      1      0      1       CONTINUA
42          107225208    202104      1      0      1       CONTINUA
43           58953650    202004      1      0      1       CONTINUA
44           73397765    201908      1      0      1       CONTINUA
45           38072385    202101      1      0      1       CONTINUA
46           38790790    201907      1      0      1       CONTINUA
47          170826206    202010      1      0      1       CONTINUA
48           34435781    201908      1      0      1       CONTINUA
49           34671025    202011      1      0      1       CONTINUA
50           69537284    201908      1      0      1       CONTINUA
51          101774714    202009      1      0      1       CONTINUA
52           84502102    202001      1      0      1       CONTINUA
53           35879201    201906      1      0      1       CONTINUA
54           50029055    201903      1      0      1       CONTINUA
55          127332799    202104      1      0      1       CONTINUA
56           88649808    202011      1      0      1       CONTINUA
57           78751667    201911      1      0      1       C

In [30]:
%%sql
SELECT 
    * 
FROM targets
WHERE numero_de_cliente=30147596

numero_de_cliente  foto_mes  mes_0  mes_1  mes_2 clase_ternaria
0            30147596    201901      1      1      1       CONTINUA
1            30147596    201902      1      1      1       CONTINUA
2            30147596    201903      1      1      1       CONTINUA
3            30147596    201904      1      1      1       CONTINUA
4            30147596    201905      1      1      1       CONTINUA
5            30147596    201906      1      1      1       CONTINUA
6            30147596    201907      1      1      1       CONTINUA
7            30147596    201908      1      1      1       CONTINUA
8            30147596    201909      1      1      1       CONTINUA
9            30147596    201910      1      1      1       CONTINUA
10           30147596    201911      1      1      1       CONTINUA
11           30147596    201912      1      1      1       CONTINUA
12           30147596    202001      1      1      0         BAJA+2
13           30147596    202002      1      0      0         BAJA+1

In [5]:
%%sql
alter table competencia_02 add column clase_ternaria VARCHAR(10)

Empty DataFrame
Columns: [Success]
Index: []

In [6]:
%%sql
update competencia_02
set clase_ternaria = targets.clase_ternaria
from targets
where competencia_02.numero_de_cliente = targets.numero_de_cliente and competencia_02.foto_mes = targets.foto_mes;

Empty DataFrame
Columns: [Success]
Index: []

Cuento cuántos casos tengo con BAJA+2 y cuántas filas tiene mi periodo a predecir 202107:

In [ ]:
%%sql
SELECT 
    foto_mes, 
    clase_ternaria, 
    count(*) as cant_bajas
FROM competencia_02
WHERE clase_ternaria = 'BAJA+2'
GROUP BY 
    foto_mes, 
    clase_ternaria
ORDER BY foto_mes DESC

In [13]:
%%sql
SELECT 
    foto_mes, 
    clase_ternaria, 
    count(*) as cant_bajas
FROM competencia_02
WHERE clase_ternaria IS NULL 
GROUP BY 
    foto_mes, 
    clase_ternaria
ORDER BY foto_mes DESC

foto_mes clase_ternaria  cant_bajas
0    202107            NaN      164682
1    202106            NaN      164420

Guardamos el dataset final:

In [14]:
%%sql
copy competencia_02 to 'C:/Users/vanes/Documents/UBA/2do_cuatrimestre/DMEyF/datasets/competencia_02.csv.gz' (FORMAT CSV, HEADER)

Empty DataFrame
Columns: [Success]
Index: []